File: solution-python/027_remove_element.py
"""
	027. Remove Element
	@name: li jin
	@date: Mar 03, 2016
	@link: https://leetcode.com/problems/remove-element/
	@time: 46 ms
"""

class Solution(object):
    def removeElement(self, nums, val):
        n, p = len(nums), 0
        i = 0
        while i < n - p:
            if nums[i] == val:
                p += 1
                while p < n - i and nums[-p] == val:
                    p += 1
                if p < n - i:
                    nums[i] = nums[-p]
            i += 1
        return n - p

File: solution-cpp/027_remove_element.cpp
/*
	027. Remove Element
	@name: li jin
	@date: Mar 03, 2016
	@link: https://leetcode.com/problems/remove-element/
	@time: 4 ms
*/

class Solution {
public:
    int removeElement(vector<int>& nums, int val) {
        int idx, cur;
        int n = nums.size();
        for (idx = cur = 0; idx < n - cur; idx++) {
            if (nums[idx] == val) {
                do {
                    cur++;
                } while (idx < n - cur && nums[n - cur] == val);
                if (idx < n - cur) nums[idx] = nums[n - cur];
            }
        }
        return n - cur;
    }
};
